### PR TITLE
Feature auto bed leveling

### PR DESCRIPTION
**Pull request by kevingelion**
@jminardi @riningerr please take a look!
Added M238 which contains original code of M235 but returns the value instead of printing. M235 now calls M238.
M241 is a version of G4 that can be called from anywhere

### CODE REVIEW

File: Marlin/Marlin.h [hunk 1]
```diff
@@ -110,6 +110,8 @@ FORCE_INLINE void serialprintPGM(const char *str) {
 
 void get_command();
 void process_commands();
+uint16_t gcode_M238(uint8_t power = 0);
+void gcode_M241(long num_milliseconds = NULL);
```
Comment (riningerr) [inline Marlin/Marlin.h]: The gcode functions are usually inlined and defined in Marlin_main.cpp. Any reason you want to put the defs here?

Also, you should probably put some `#ifdef` stuff around them if you are using the distance sensor or anything else that isn't in use by default.

Comment (kevingelion) [inline Marlin/Marlin.h]: This is so M238 and M241 can be called from any other function/in a different scope. For example, when probing the bed level, M241 will be called to pause for a certain amount of time.\

But yeah, definitely a good idea about the `#ifdef` stuff.

Comment (riningerr) [inline Marlin/Marlin.h]: I believe those should only called within Marlin_main.cpp though. @jminardi what are your thoughts?

File: Marlin/Marlin_main.cpp [hunk 2]
```diff
@@ -4160,31 +4230,15 @@ inline void gcode_M226() {
...
-    // (This value was just taken from M234 for simplicity)
-    if(power > ADC_SAMPLE_POWER) {
-      power = ADC_SAMPLE_POWER;
-    }
-    // Take specified amount of readings
-    for(i = 0; i < num_samples; i++) {
-      sample_sum += EXT_ADC_READ_0;
-    }
-    // Take average of sample readings
-    sample_avg = sample_sum >> power;
```
Comment (riningerr) [inline Marlin/Marlin_main.cpp]: I don't understand why you've redefined M235 as M238? It seems to be the same thing, but M235 just returns what M238 does?

I think we should leave M235 as is and call gcode_M235() if it is needed.

File: Marlin/Marlin_main.cpp [hunk 3]
```diff
@@ -4381,6 +4557,32 @@ inline void gcode_M226() {
 
 #endif // CHDK || PHOTOGRAPH_PIN
 
+/*
+* M241 - Dwell for a given amount of time in milliseconds (500 by default)
+*/
+void gcode_M241(long num_milliseconds) {
```
Comment (riningerr) [inline Marlin/Marlin_main.cpp]: Could you not use a `uint16_t` for `num_milliseconds` here? That would give you over a minute of possible dwell time. An `unsigned long` gives you over 1000 hours of possible dwell time.

File: Marlin/Marlin_main.cpp [hunk 4]
```diff
@@ -4225,6 +4279,128 @@ inline void gcode_M226() {
...
+      SERIAL_EOL;
+    }
+    enqueuecommands_P(PSTR("G28"));
+    char cmd[30] = { 0 };
+
+    sprintf_P(cmd, PSTR("M237 V%d"), verbose_level);
+    enqueuecommand(cmd);
+  }
+#endif
+
```
Comment (riningerr) [inline Marlin/Marlin_main.cpp]: I'm not sure we need a function that combines G28 and M237 when we can just issue both of those commands.

File: Marlin/Marlin_main.cpp (hunk 2 above)
Comment (kevingelion) [inline Marlin/Marlin_main.cpp]: `M235()` only prints data - we need the values returned and called in a different scope

File: Marlin/Marlin.h (hunk 1 above)
Comment (jminardi) [inline Marlin/Marlin.h]: I dont think it hurts to put them in the .h file. Please do add ifdefs though

File: Marlin/Marlin_main.cpp (hunk 4 above)
Comment (jminardi) [inline Marlin/Marlin_main.cpp]: I think we should leave it as it is a work around for a bug

**Comment by jminardi** [comment]
Please test this on one of the gen3b printers. If it works as expected then feel free to merge

**Comment by riningerr** [comment]
Closed because development continued on a different branch (Kevin_Integration) for Gen3c

**Pull request closed without merging**